src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -62,6 +64,68 @@ static void check(Coordination::Error code, const std::string & path)
         throw KeeperException::fromPath(code, path);
 }
 
+bool ZooKeeperAvailabilityZoneMap::needTryOtherHost(const std::string &local_az, const std::set<std::string> &attempted_host)
+{
+    std::lock_guard lock(mutex);
+    for (const auto &host_az : az_by_host)
+    {
+        const auto & host = host_az.first;
+        const auto & az = host_az.second;
+        if ((az == AZ_UNKNWON || az == local_az) && !attempted_host.contains(host))
+            return true;
+    }
+    return false;
+}
+
+std::pair<std::string, bool> parseForSocketAddress(const std::string& raw_host)
+{
+    std::pair<std::string, bool> result;
+    bool secure = startsWith(raw_host, "secure://");
+    if (secure)
+        result.first = raw_host.substr(strlen("secure://"));
+    else
+        result.first = raw_host;
+    result.second = secure;
+    return result;
+}
+
+bool isKeeperHostDNSAvailable(Poco::Logger* log, const std::string& host, bool& dns_error)
+{
+    auto address = parseForSocketAddress(host).first;
+    /// We want to resolve all hosts without DNS cache for keeper connection.
+    Coordination::DNSResolver::instance().removeHostFromCache(address);
+    try
+    {
+        const Poco::Net::SocketAddress host_socket_addr{address};
+    }
+    catch (const Poco::Net::HostNotFoundException & e)
+    {
+        /// Most likely it's misconfiguration and wrong hostname was specified
+        LOG_ERROR(log, "Cannot use ZooKeeper host {}, reason: {}", host, e.displayText());
+        return false;
+    }
+    catch (const Poco::Net::DNSException & e)
+    {
+        /// Most likely DNS is not available now
+        dns_error = true;
+        LOG_ERROR(log, "Cannot use ZooKeeper host {} due to DNS error: {}", host, e.displayText());
+        return false;
+    }
+    // NOTE: check other exception case how it's handled before.
```

Comment:
Check this.(to myself)

Reply:
https://github.com/ClickHouse/ClickHouse/blob/c465933b1d0173422df6a9fadaec4ba2f5611f17/src/Common/ZooKeeper/ZooKeeper.cpp#L103
same only these two exceptions.